operator/cmd/mesh/manifest-generate.go

```diff
@@ -124,111 +114,90 @@ func ManifestGenerateCmd(ctx cli.Context, rootArgs *RootArgs, mgArgs *ManifestGe
-	var rawOutput []string
-	var output []string
-	for _, mfs := range mm {
-		rawOutput = append(rawOutput, mfs...)
-	}
-	objects, err := object.ParseK8sObjectsFromYAMLManifest(strings.Join(rawOutput, helm.YAMLSeparator))
-	if err != nil {
-		return nil, err
+func sortManifests(raw []manifest.ManifestSet) []string {
+	all := []manifest.Manifest{}
+	for _, m := range raw {
+		all = append(all, m.Manifests...)
 	}
-	// For a given group of objects, sort in order to avoid missing dependencies, such as creating CRDs first
-	objects.Sort(object.DefaultObjectOrder())
-	for _, obj := range objects {
-		yml, err := obj.YAML()
-		if err != nil {
-			return nil, err
+	slices.SortStableFunc(all, func(a, b manifest.Manifest) int {
+		if r := cmp.Compare(objectKindOrdering(a), objectKindOrdering(b)); r != 0 {
+			return r
 		}
-		output = append(output, string(yml))
-	}
-
-	return output, nil
-}
-
-// RenderToDir writes manifests to a local filesystem directory tree.
-func RenderToDir(manifests name.ManifestMap, outputDir string, dryRun bool, l clog.Logger) error {
-	l.LogAndPrintf("Component dependencies tree: \n%s", helmreconciler.InstallTreeString())
-	l.LogAndPrintf("Rendering manifests to output dir %s", outputDir)
-	return renderRecursive(manifests, helmreconciler.InstallTree, outputDir, dryRun, l)
-}
-
-func renderRecursive(manifests name.ManifestMap, installTree helmreconciler.ComponentTree, outputDir string, dryRun bool, l clog.Logger) error {
-	for k, v := range installTree {
-		componentName := string(k)
-		// In cases (like gateways) where multiple instances can exist, concatenate the manifests and apply as one.
-		ym := strings.Join(manifests[k], helm.YAMLSeparator)
-		l.LogAndPrintf("Rendering: %s", componentName)
-		dirName := filepath.Join(outputDir, componentName)
-		if !dryRun {
-			if err := os.MkdirAll(dirName, os.ModePerm); err != nil {
-				return fmt.Errorf("could not create directory %s; %s", outputDir, err)
-			}
+		if r := cmp.Compare(a.GroupVersionKind().Group, b.GroupVersionKind().Group); r != 0 {
+			return r
 		}
-		fname := filepath.Join(dirName, componentName) + ".yaml"
-		l.LogAndPrintf("Writing manifest to %s", fname)
-		if !dryRun {
-			if err := os.WriteFile(fname, []byte(ym), 0o644); err != nil {
-				return fmt.Errorf("could not write manifest config; %s", err)
-			}
+		if r := cmp.Compare(a.GroupVersionKind().Kind, b.GroupVersionKind().Kind); r != 0 {
+			return r
 		}
+		return cmp.Compare(a.GetName(), b.GetName())
+	})
+	return slices.Map(all, func(e manifest.Manifest) string {
+		// marshal the object instead of using the raw content to normalized the output
+		// This is likely not good behavior
+		res, _ := yaml.Marshal(e.Object)
+		return string(res)
+	})
+}
 
-		kt, ok := v.(helmreconciler.ComponentTree)
-		if !ok {
-			// Leaf
-			return nil
-		}
-		if err := renderRecursive(manifests, kt, dirName, dryRun, l); err != nil {
-			return err
-		}
+func objectKindOrdering(m manifest.Manifest) int {
```

Comment:
Can we not simply use `helm//pkg/releaseutil` here so that we get _exactly_ the same ordering semantics as helm?

We should[ follow their ordering for install and uninstall](https://github.com/helm/helm/blob/main/pkg/releaseutil/kind_sorter.go#L31) as well - xref: https://github.com/istio/istio/issues/51998

("that would be great as a followup" is as always a reasonable answer)

Reply:
My answer is either "that would be a great as a followup" or "no, for <reason I forgot>".

I think the reason I didn't do it was I didn't want tests results to change which makes it hard to review, but its possible I found another issue I cannot recall...

Reply:
Oh there is one minor quirk that removing the clusterroles/service account before the deployment means we cannot release leader election locks, which means new Istiods are not full functional for ~45s. See #52424. This doesn't matter much beyond our tests where it really matters to be able to uninstall + reinstall quickly though

Reply:
Ah ok, that's good to know.